[MRG] Calculation of information in nats or bits depending on log_base param

HJLebbink opened the pull request:
#### Reference Issues https://github.com/scikit-learn/scikit-learn/issues/10778

#### What does this implement/fix?
Added a parameter log_base to toggle between calculation of information in bits vs nats.
Currently, the calculation of information is only for nats, which is not conventional and confusing. Log_base defaults to 'e' such that existing code will not break. Providing log_base = 2 will give the results in bits units.

#### Any other comments?
There is an SO thread about this [here](https://stackoverflow.com/questions/24686374/pythons-implementation-of-mutual-information)

HJLebbink committed:
Calculation of information in nats or bits depending on log_base param

HJLebbink committed:
Calculation of information in nats or bits depending on log_base param

HJLebbink commented:
Sorry travis, has been a while since i SAW a punch card...

HJLebbink committed:
PEP8 update

HJLebbink committed:
Merge remote-tracking branch 'origin/master' into option-for-nats-vs-bits

# Conflicts:
#	sklearn/metrics/cluster/supervised.py

HJLebbink committed:
tyring to please PEP8

jnothman commented:
please add a test of the parameter and of the warning. Thanks

qinhanmin2014 reviewed (commented):
Please fix the flake8 errors to make CIs green :)
(See https://travis-ci.org/scikit-learn/scikit-learn/jobs/352521341)

HJLebbink commented:
@qinhanmin2014 I've just wasted too long trying to understand how to format 
`log_outer = -log(outer, log_base) + log(pi.sum(), log_base)
                + log(pj.sum(), log_base)`

Im almost considering to abandon this...

qinhanmin2014 commented:
Simply split the line into multiple lines and make sure that each line does not exceed 79 characters.
(You can use flake8 file_name.py to check locally)

----- Original Message -----
From: Henk-Jan Lebbink <notifications@github.com>
To: scikit-learn/scikit-learn <scikit-learn@noreply.github.com>
Cc: Hanmin Qin <qinhanmin2005@sina.com>,  Mention <mention@noreply.github.com>
Subject: Re: [scikit-learn/scikit-learn] Calculation of information in nats or bits depending on log_base param (#10799)
Date: 2018-03-13 17:34

@qinhanmin2014 I've just wasted too long trying to understand how to format

log_outer = -log(outer, log_base) + log(pi.sum(), log_base) + log(pj.sum(), log_base)
Im almost considering to abandon this...

&mdash;
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

HJLebbink committed:
test updated to explicitly return results in nats.

HJLebbink commented:
Help needed: I do not know how to make the requested test in [test_supervised](https://github.com/HJLebbink/scikit-learn/blob/9f7e4bbb23681dae94353f0b041a90a64c3eb421/sklearn/metrics/cluster/tests/test_supervised.py#L33) to check a FutureWarning in [supervised](https://github.com/HJLebbink/scikit-learn/blob/9f7e4bbb23681dae94353f0b041a90a64c3eb421/sklearn/metrics/cluster/supervised.py#L109).

HJLebbink committed:
Merge remote-tracking branch 'scikit-learn/master' into option-for-nats-vs-bits

jnothman commented:
search for other uses of assert_warns in the repo

HJLebbink committed:
added tests for: 1] future warnings, 2] log_base param

HJLebbink committed:
test fix: test invalid param for entropy

HJLebbink committed:
cleanup

HJLebbink committed:
added support for base 10 log (dits)

HJLebbink committed:
added log10 support to expected_mutual_info

HJLebbink commented:
Question: should I mention in de documentation of the parameter log_base (eg [here])(https://github.com/HJLebbink/scikit-learn/blob/f7b13015a65efd98085f0263c8bced684b2394a9/sklearn/metrics/cluster/supervised.py#L920) that the default is changed in version 0.22.

Related: how do we make sure we will not forget to actually change the default assignment of log_base for version 0.22?

jnothman commented:
We do changes when the release cycle comes around. Grep is our friend.

yes, can note that it will change in the docstring. Thanks

On 15 March 2018 at 21:09, Henk-Jan Lebbink <notifications@github.com>
wrote:

> Question: should I mention in de documentation of the parameter log_base
> (eg [here])(https://github.com/HJLebbink/scikit-learn/blob/
> f7b13015a65efd98085f0263c8bced684b2394a9/sklearn/metrics/
> cluster/supervised.py#L920) that the default is changed in version 0.22.
>
> Related: how do we make sure we will not forget to actually change the
> default assignment of log_base for version 0.22?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/10799#issuecomment-373324805>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66eRAsBMkG_VVK-XQ7ML0q1XwhlCks5tej28gaJpZM4SmsBB>
> .
>

HJLebbink committed:
documentation updated

HJLebbink commented:
Is there anything I still need to do on for this PR?

HJLebbink commented:
I like to wrap this up, is there anything I can do?

jnothman commented on sklearn/metrics/cluster/tests/test_supervised.py:
```diff
@@ -56,6 +67,31 @@ def test_perfect_matches():
+        normalized_mutual_info_score
+    ]
+    for score_func in score_funcs:
+        assert_warns(FutureWarning, score_func, [0, 0], [0, 1])
```
Please also check the return value

jnothman reviewed (commented):
Thanks for the ping. Are there also uses in doc/ that need updating?

I'm a little consented about a warning raised with every call to log. Could we do the validation in the public functions instead?

Otherwise this looks good

HJLebbink committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into option-for-nats-vs-bits

HJLebbink committed:
added expected future warning in test

HJLebbink commented:
> Are there also uses in doc/ that need updating?
I updated the docs somewhat. I mentioned that functionally will change in the future.

> I'm a little consented about a warning raised with every call to log. Could we do the validation in the public functions instead?

Because the FutureWarning is only issued once (even when log is called more than once) I figured I can be placed as 'deep' as possible just to prevent code duplication, especially because someone will need to grep/replace this code somewhere in the future. The log method is not exported and only used by the three methods that I updated. 

But it is up to you where we do the validation (since you will be grebbing...)

jnothman commented:
Tests failing. No changes to doc/?

HJLebbink committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into option-for-nats-vs-bits

lorentzenchr commented:
See #14576.
